metronome/checkpointing/service/src/io/iohk/metronome/checkpointing/service/CheckpointingService.scala

```diff
@@ -0,0 +1,208 @@
+package io.iohk.metronome.checkpointing.service
+
+import cats.Applicative
+import cats.data.{NonEmptyList, NonEmptyVector, OptionT}
+import cats.effect.concurrent.{Ref, Semaphore}
+import cats.effect.{Concurrent, Resource, Sync}
+import cats.implicits._
+import io.iohk.metronome.checkpointing.CheckpointingAgreement
+import io.iohk.metronome.checkpointing.models.{
+  Block,
+  CheckpointCertificate,
+  Ledger
+}
+import io.iohk.metronome.checkpointing.service.storage.LedgerStorage
+import io.iohk.metronome.crypto.ECPublicKey
+import io.iohk.metronome.hotstuff.consensus.basic.{Phase, QuorumCertificate}
+import io.iohk.metronome.hotstuff.service.ApplicationService
+import io.iohk.metronome.hotstuff.service.storage.{
+  BlockStorage,
+  ViewStateStorage
+}
+import io.iohk.metronome.storage.KVStoreRunner
+
+class CheckpointingService[F[_]: Sync: Applicative, N] private (
+    ledgerTree: Ref[F, Map[Block.Hash, (Ledger, Block.Header)]],
+    lastExecutedHeader: Ref[F, Block.Header],
+    ledgerUpdateSemaphore: Semaphore[F],
+    maxConcurrentUpdates: Int,
+    viewStateStorage: ViewStateStorage[N, CheckpointingAgreement],
+    ledgerStorage: LedgerStorage[N],
+    blockStorage: BlockStorage[N, CheckpointingAgreement]
+)(implicit storeRunner: KVStoreRunner[F, N])
+    extends ApplicationService[F, CheckpointingAgreement] {
+
+  override def createBlock(
+      highQC: QuorumCertificate[CheckpointingAgreement]
+  ): F[Option[Block]] = ???
+
+  override def validateBlock(block: Block): F[Option[Boolean]] = {
+    for {
+      lastExec <- lastExecutedHeader.get
```

Comment:
This should probably be within the semaphore: in case a commit moves the last executed header, this value that we read before trying to take out the semaphore could be obsolete.

Reply:
No longer relevant